Topics/Software_Tools/Git/Git_Merging_and_Rebasing.md

```diff
@@ -11,4 +11,118 @@ Both `git merge` and `git rebase` are methods for integrating new commits that a
+
+Rebasing comes with a big advantage though: it creates an entirely linear history after the rebase because it acts as if the branch that is being rebased was created after the latest commit of the branch being rebased on. This may not seem like such a big deal, but when you’re trying to track down what commit caused a certain bug, you will be wishing that your git history was linear!
+
+Doesn't this:
+> A -> B -> C -> D -> E -> F -> G
+
+Look better than this?
+>
+> A -> B -> C -> E -> F -> (merge commit)
+>
+>             ↳ D -> G -> ⬏
+
+
+
+### A small warning
+
+You should take care with rebasing if you’re working with a remote repository! Rebasing is rewriting history, and as a general rule, rewriting history is dangerous when done on a branch that multiple people work on. However, if you are the only person working on that branch, rebasing is harmless if you know what you are doing.
+
+The general advice is that you should only rebase branches that you are working on by yourself. If you expect multiple people to be working on a branch, you should stick to merging. You may also find that when two branches have diverged for a long time, rebasing may cause you to need to solve many rounds of merge conflicts, while merging requires only 1 round of merge conflicts (remember the merge commit?).
+
+## `git rebase --interactive`
+
+While git rebase is used for rebasing a branch on another branch, we can also edit the commits before they are applied using the --interactive flag. In most cases though, we use `git rebase --interactive` on the currently checked out branch.
+
+This sounds a bit scary, but what it essentially boils down to is that you can rewrite history at will! Want to squash two commits into one? Or maybe drop a commit that you made two months ago? Or maybe swap the order of two commits? Interactive rebase gives you all these powers.
+
+In this section, we’ll cover the most common operations: pick, reword, edit, squash, fixup, and drop.
+
+To start an interactive rebase, use `git rebase --i <commit hash>`, where <commit hash> is the last commit hash you DON’T want to modify (i.e. your base commit).
+
+When you start an interactive rebase, you’ll be met with a log of the commits you’ll be rebasing (oldest commits at the top).
+
+By default, each commit has “pick” to its left, which means Git will simply add that commit as-is.
+
+- To change the order of your commits, copy-paste your commits into the order you want.
+- To change a commit’s message, replace “pick” with “reword”.
+- To squash a commit into the previous commit, replace “pick” with “squash”.
+- To squash a commit into the previous commit but use the previous commit’s message, replace “pick” with “fixup”.
+- To drop (i.e. delete) a commit, replace “pick” with “drop”.
+- To edit or split up a commit, replace “pick” with “edit”.
+
```

Comment:
It might be nice here to include an image of what the CLI looks like, so that you can see where each action word should go.

Reply:
Added a small running example to each action word.